cmd/stencil/module_tst.go

```diff
@@ -0,0 +1,213 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/jaredallard/cmdexec"
+	"github.com/urfave/cli/v3"
+	"go.rgst.io/stencil/v2/internal/cmd/stencil"
+	"go.rgst.io/stencil/v2/internal/yaml"
+	"go.rgst.io/stencil/v2/pkg/configuration"
+	"go.rgst.io/stencil/v2/pkg/slogext"
+)
+
+// Test represents a stencil test.
+type Test struct {
+	// Name is the name of the test
+	Name string
+
+	// Dir is the directory where the test lives
+	Dir string
+
+	// Failed denotes if the test failed or not
+	Failed bool
+
+	// FailureReason is the optional reason that a test failed
+	FailureReason error
+
+	// TemplateRepoManifest is the manifest of the template repository
+	// this test was testing.
+	TemplateRepoManifest *configuration.TemplateRepositoryManifest
+}
+
+// runTest runs a test in the provided directory with the provided test
+// name
+func runTest(ctx context.Context, log slogext.Logger, dir string, t *Test) error {
+	tmpTestDir, err := os.MkdirTemp("", "stencil-test-*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory for test: %w", err)
+	}
+
+	if err := os.Chdir(tmpTestDir); err != nil {
+		return fmt.Errorf("failed to change working directory: %w", err)
+	}
+
+	// Restore original work dir and clean up the temporary test directory
+	defer func() {
+		if err := os.Chdir(dir); err != nil {
+			log.WithError(err).Warn("failed to restore previous working directory")
+		}
+
+		if err := os.RemoveAll(tmpTestDir); err != nil {
+			log.WithError(err).Warn("failed to clean up temporary working directory")
+		}
+	}()
+
+	manifestPath := filepath.Join(t.Dir, "stencil.yaml")
+	serializedMfPath := filepath.Join(tmpTestDir, "stencil.yaml")
+
+	mf, err := configuration.LoadManifest(manifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to load stencil.yaml: %w", err)
+	}
+
+	// Ensure we're using a name that matches the test name and that we're
+	// using a local version of the module that we're trying to test.
+	mf.Name = t.Name
+
+	// TODO(jaredallard): Allow including other modules, but ensure that
+	// we always control the version of the tested module
+	mf.Modules = []*configuration.TemplateRepository{
+		{Name: t.TemplateRepoManifest.Name, Version: "=0.0.0"},
+	}
+
+	mf.Replacements = map[string]string{
+		t.TemplateRepoManifest.Name: dir,
+	}
+
+	b, err := yaml.Marshal(mf)
+	if err != nil {
+		return fmt.Errorf("failed to marshal manifest modifications: %w", err)
+	}
+
+	if err := os.WriteFile(serializedMfPath, b, 0o600); err != nil {
+		return fmt.Errorf("failed to write manifest modifications to temp dir: %w", err)
+	}
+
+	// TODO(jaredallard): plumb in logger from newmoduletestcommand
+	cmd := stencil.NewCommand(slogext.NewNullLogger(), mf, false, false)
```

Comment:
because you're doing the replacements by hand, people are going to be running this test stuff to iterate on their "unit test", since you can't actually run stencil in the test dir and have it work (because replacements won't be in the file unless you manually add it), so i'd expect the logger here to dump any details if the test fails